File: library/dkv.py
import subprocess
import os

class DKVClient:
    def __init__(self, ctrler_addr="localhost:9000", client_bin="dkv-client"):
        """
        Initializes the DKV client wrapper.
        
        :param ctrler_addr: Address of the Shard Controller or config store.
        :param client_bin: Path to the compiled dkv-client CLI binary.
        """
        self.ctrler_addr = ctrler_addr
        self.client_bin = client_bin

    def put(self, key: str, value: str, version: int = None) -> bool:
        """
        Writes a key-value pair to the distributed key-value store.
        
        :param key: Key to write.
        :param value: Value to write.
        :param version: Optional version constraint.
        :return: True if successful, raises Exception otherwise.
        """
        cmd = [self.client_bin, "--ctrler-addr", self.ctrler_addr, "put", key, value]
        if version is not None:
            cmd.append(str(version))
        
        result = subprocess.run(cmd, capture_output=True, text=True)
        if result.returncode != 0:
            raise Exception(f"DKV Put failed: {result.stderr.strip()}")
        return result.stdout.strip() == "OK"

    def get(self, key: str) -> tuple[str, int]:
        """
        Reads a key's value and version from the distributed store.
        
        :param key: Key to read.
        :return: A tuple of (value, version). If key is not found, returns (None, 0).
        """
        cmd = [self.client_bin, "--ctrler-addr", self.ctrler_addr, "get", key]
        result = subprocess.run(cmd, capture_output=True, text=True)
        if result.returncode != 0:
            stderr = result.stderr.strip()
            if "Key not found" in stderr or result.returncode == 1:
                return None, 0
            raise Exception(f"DKV Get failed: {stderr}")
        
        output = result.stdout.strip()
        if output.startswith("Value:"):
            try:
                # Expected format: "Value: <val> (Version: <ver>)"
                val_part, ver_part = output.split(" (Version: ")
                val = val_part[len("Value: "):]
                ver = int(ver_part[:-1])
                return val, ver
            except Exception:
                raise Exception(f"Failed to parse DKV client output: {output}")
        return None, 0

File: library/dkv.ts
import { execFile } from 'child_process';
import { promisify } from 'util';

const execFileAsync = promisify(execFile);

export class DKVClient {
    private ctrlerAddr: string;
    private clientBin: string;

    /**
     * Initializes the DKV client wrapper.
     * 
     * @param ctrlerAddr Address of the Shard Controller or config store.
     * @param clientBin Path to the compiled dkv-client CLI binary.
     */
    constructor(ctrlerAddr = 'localhost:9000', clientBin = 'dkv-client') {
        this.ctrlerAddr = ctrlerAddr;
        this.clientBin = clientBin;
    }

    /**
     * Writes a key-value pair to the distributed key-value store.
     * 
     * @param key Key to write.
     * @param value Value to write.
     * @param version Optional version constraint.
     * @returns Promise resolving to true if successful, rejects on failure.
     */
    async put(key: string, value: string, version?: number): Promise<boolean> {
        const args = ['--ctrler-addr', this.ctrlerAddr, 'put', key, value];
        if (version !== undefined) {
            args.push(version.toString());
        }

        try {
            const { stdout } = await execFileAsync(this.clientBin, args);
            return stdout.trim() === 'OK';
        } catch (error: any) {
            throw new Error(`DKV Put failed: ${error.stderr || error.message}`);
        }
    }

    /**
     * Reads a key's value and version from the distributed store.
     * 
     * @param key Key to read.
     * @returns Promise resolving to an object containing value and version. Returns value null and version 0 if key not found.
     */
    async get(key: string): Promise<{ value: string | null; version: number }> {
        const args = ['--ctrler-addr', this.ctrlerAddr, 'get', key];

        try {
            const { stdout } = await execFileAsync(this.clientBin, args);
            const output = stdout.trim();
            if (output.startsWith('Value:')) {
                const valPart = output.substring('Value: '.length);
                const verSplit = valPart.split(' (Version: ');
                const val = verSplit[0];
                const ver = parseInt(verSplit[1].slice(0, -1), 10);
                return { value: val, version: ver };
            }
            return { value: null, version: 0 };
        } catch (error: any) {
            if (error.code === 1 || (error.stderr && error.stderr.includes('Key not found'))) {
                return { value: null, version: 0 };
            }
            throw new Error(`DKV Get failed: ${error.stderr || error.message}`);
        }
    }
}
